src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/SocketInput.cs

```diff
@@ -103,9 +103,9 @@ public void IncomingComplete(int count, Exception error)
                 {
                     FinReceived();
                 }
-
-                Complete();
             }
+
+            Complete();
```

Comment:
The call to Complete() here (or at the very least the Interlocked.Exchange call inside), needs to share the same lock as:

                    if (!examined.IsDefault &&
                        examined.IsEnd &&
                        ReadingInput)
                    {
                        _manualResetEvent.Reset();

                        Interlocked.CompareExchange(
                            ref _awaitableState,
                            _awaitableIsNotCompleted,
                            _awaitableIsCompleted);
                    }

in ConsumingComplete(...). Otherwise, the consumer can get stuck forever awaiting data at the end of the response that was already there.